06x-new-kafka-roller.md

```diff
@@ -0,0 +1,326 @@
+| maxRestartParallelism  | 1             | Yes             | The maximum number of broker nodes that can be restarted in parallel. This will be exposed to the user via the new environment variable `STRIMZI_MAX_RESTART_BATCH_SIZE`. However, if there are multiple brokers in `NOT_RUNNING` state, they may get restarted in parallel despite this configuration for a faster recovery.    
+| postRestartDelay  | 0             | Yes             | Delay between restarts of nodes or batches. It's set to 0 by default, but can be adjusted by users to slow down the restarts. This will also help JIT to reach a steady state and to reduce impact on clients.   
+| restartAndPreferredLeaderElectionDelay  | 10 seconds             | No             | Delay between restart and triggering partition leader election so that just-rolled broker is leading all the partitions it is the preferred leader for. This is to avoid situations where leaders moving to a newly started node that does not yet have established networking to some outside networks, e.g. through load balancers.
+
+### Algorithm
+
+1. **Initialize Context for Each Node:**
+   Create a context object with the following data:
+   ```
+   Context: {
+       nodeRef: <NodeRef from KafkaReconciler>,
+       nodeRoles: <Set using pod labels `strimzi.io/controller-role` and `strimzi.io/broker-role`>,
+       state: UNKNOWN,
+       lastTransition: <SYSTEM_TIME>,
+       restartReason: <Result of predicate function from KafkaReconciler>,
+       numRestartAttempts: 0,
+       numReconfigAttempts: 0,
+       numRetries: 0
+   }
+   ```
+   Contexts are recreated in each reconciliation with the above initial data.
+
+2. **Transition Node States:**
+   Update each node's state based on information from abstracted sources. If failed to retrieve information, the current reconciliation immediately fails. When the next reconciliation is triggered, it will restart from step 1.
+
+3. **Handle `NOT_READY` Nodes:**
+   Wait for `NOT_READY` nodes to become `READY` within `operationTimeoutMs`.
+
+4. **Categorize Nodes:**
+   Group nodes based on their state and connectivity:
+   - `RESTART_NOT_RUNNING`: Nodes in `NOT_READY` state.
+   - `WAIT_FOR_LOG_RECOVERY`: Nodes in `RECOVERING` state.
+   - `RESTART_UNRESPONSIVE`: Nodes unresponsive via Admin API.
+   - `MAYBE_RECONFIGURE_OR_RESTART`: Broker nodes with empty reason lists and no previous restarts/reconfigurations.
+   - `RESTART`: Nodes with reasons for restart and no previous restarts.
+   - `NOP`: Nodes needing no operation.
+
+5. **Wait for Log Recovery:**
+   Wait for `WAIT_FOR_LOG_RECOVERY` nodes to become `READY` within `operationTimeoutMs`. If timeout is reached and `numRetries` exceeds `maxRetries`, throw `UnrestartableNodesException`. Otherwise, increment `numRetries` and repeat from step 2.
```

Comment:
For all these steps I think it would be really valuable to explain the _why_. Here we're willing to wait for brokers in log recovery because the following steps will result in actions, like restarting other brokers, which will be directly visible to clients. We prefer to start from a cluster that's as close to fully functional as possible.

Reply:
We also need to explain why are we willing to wait for log recovery here, but not willing to wait for all the brokers replicas to rejoin the ISR? 

IIRC the reason we had was the KafkaRoller's job was to restart things, but we didn't want to give it any responsibility for throttling interbroker replication, and we can't guarantee (for all possible workloads) that brokers will always be able to catch up to the LEO (within a reasonable time).

Reply:
Thanks @tombentley. I have added the reasons.